### PR TITLE
Make sure dispatch can be called multiple times per-tick

### LINKED ISSUE #1206: Bulk navigation action dispatch
Sometimes it is necessary to dispatch more than one type of navigation action at once. In my case I found a situation where I needed to trigger a `setParams` and `navigate` action at the same time (navigating to a new scene at the same time as updating the current scene's params to close the menu with the button that triggered the navigation).

However you can't actually just do `navigation.setParams(...); navigation.navigate('...');` because from what I can tell dispatches trigger a setState and you don't get the new state you need till the next tick.

So I believe it makes sense for us to have way to trigger multiple actions at the same time, at least at the low level something like:

```js
navigation.dispatch(NavigationActions.bulk({
  actions: [
    NavigationActions.setParams({
      key: navigation.state.key,
      params: {...},
    }),
    NavigationActions.navigate({
      routeName: '...'
    })
  ]
}));

### PR DESCRIPTION
**Pull request by dantman**
Mostly fixes #1206

## Test plan (required)

I've added tests:
* To validate the dispatch => true/false return value behaviour is retained
* `dispatch` functions normally, updating the state by the next tick
* The change to `dispatch` works, allowing 2 calls to dispatch to work in the same tick without throwing away data
  * react-test-runner applies setState synchronously instad of waiting for the next tick like React normally does so the test file makes sure to inject a batching strategy that applies setState when `jest.runOnlyPendingTimers();` is called
  * The test also verifies that state has **not** been updated synchronously immediately after the dispatch call to ensure that  jest committing setState synchronously does not result in a falsely passing test

### CODE REVIEW

**Comment by dantman** [comment]
Test failure appears to be due to pre-existing issues with the flow types

File: src/createNavigationContainer.js [hunk 1]
```diff
@@ -152,6 +152,13 @@ export default function createNavigationContainer<T: *>(
       this._validateProps(nextProps);
     }
 
+    componentDidUpdate() {
+      // Clear cached _nav every tick
+      if (this._nav === this.state.nav) {
+        delete this._nav;
```
Comment (matthamil) [inline src/createNavigationContainer.js]: I'm not sure if using `delete` is the right approach here.

Comment (dantman) [inline src/createNavigationContainer.js]: What do you think the "right" approach is then?

Comment (matthamil) [inline src/createNavigationContainer.js]: @dantman Would it not be better to set `this._nav` to `null` when you are clearing the cache? All `delete` is doing is removing the IIRC `delete` interferes with some JS engine optimizations.

Comment (dantman) [inline src/createNavigationContainer.js]: I can make the change. But this is going to be a pointless level of micro-optimization on a class like NavigationContainer; and given the React+Babel weight I have strong doubts that this class' instances are "hot" objects that use that one type of optimization.

**Comment by dantman** [comment]
CC: @ericvicenti

**Comment by Jacse** [comment]
What's the status of this, @dantman?

**Comment by dantman** [comment]
@Jacse Waiting for anyone to re-review it. I made all the requested changes and since then no-one has bothered to review it.

**Comment by aroth** [comment]
Looking forward to this as well.

**Comment by rpopovici** [comment]
@dantman I think I have a better solution to this problem. Your solution can handle two dispatches fired in a row. In my solution I am not relying on componentDidUpdate to detect when setState has completed. I am queuing all action request and dispatch them on setState callback. This way we can handle multiple dispatches fired in a row without worrying about losing data and besides this I also fixed the Transitioner in the same way. The Transitioner was able to queue only one transition but even that was flawed with reset action followed by a setParams for example..

**Comment by dantman** [comment]
@rpopovici 
* This solution isn't limited to 2 dispatches in a row, it can handle an infinite number of them.
* This solution does not rely on `componentDidUpdate` to work, the `componentDidUpdate` is just a bit of housekeeping. The code would still work even if you deleted that, all the logic is in `dispatch`.

**Comment by rpopovici** [comment]
@dantman my bad, now I understand how it works. Still, I don't think it's a good way to do it like this because you are not waiting for setState to finish. setState is async per React docs https://facebook.github.io/react/docs/react-component.html#setstate and it should not be called in a synchronous manner even if you can predict the next state. You are assuming that everything will work correctly, but you don't know how React is going to change in the future and this approach might not work anymore

**Comment by dantman** [comment]
@rpopovici
Firstly a side note: IMHO #2400 doesn't do it right either, waiting for setState like that is problematic in 2 ways.
* Firstly you can already write `dispatch + wait a tick + dispatch` on your own (setImmediate), when a developer calls `dispatch + dispatch` intentionally in the same tick all dispatches should be resolved in the same tick, not run one by one waiting for a tick to pass and transition to finish.
* Secondly you double the react updates (the setState callback is "guaranteed to fire after the update has been applied", so by waiting for it and making another setState call you now guarantee that even if you write multiple dispatches in the same tick, every single one of them will trigger a separate renderer update by React instead of allowing react to optimize those calls into one update. 

> Still, I don't think it's a good way to do it like this because you are not waiting for setState to finish. setState is async per React docs https://facebook.github.io/react/docs/react-component.html#setstate and it should not be called in a synchronous manner even if you can predict the next state. You are assuming that everything will work correctly, but you don't know how React is going to change in the future and this approach might not work anymore

React isn't going to break this, if they did it would fundamentally break react in normal usage.

The documentation you quote explicitly states:
> Subsequent calls will override values from previous calls in the same cycle, so the quantity will only be incremented once.

The async nature of setState only means that you cannot access changes in `this.state` until the next tick and need either a 2nd source of truth or the updater function.

If react were to change this, it would mean that `this.setState({a: 1}); this.setState({a: 2});` might sometime set the state to `{a: 1}` instead of`{a: 2}`. That would fundamentally break react.

And even if you were so paranoid that react would do that, I could just refactor it so all dispatches are batched into one setState call.

----

I could easily have wrote this commit without `this._nav`. In fact that is what I wrote originally looked like. And with a few lines of code I could easily change it back to that. You **do not** need a queue array, to wait a tick for setState, or have a separate queueDispatch method, you just use the updater function.

The reason I used `_nav` instead is because `getStateForAction` needs to be resolved in the same tick to return success/failure. And none of the alternatives (promise, callback, error event, etc...) were worth the breaking change and redux breakage when you can do the change without them.

**Comment by rpopovici** [comment]
@dantman OK, I will try your approach to see if it's working for my project and cancel my PR if everything works fine

**Comment by rpopovici** [comment]
@dantman you should rebase this PR against master. There are lots of changes since you open it

**Comment by dantman** [comment]
@rpopovici I tried doing that before I added that test for 5 dispatches.

Unfortunately `react-navigation` no longer runs under `react-test-renderer`. So when someone **finally** comes around to review this change I'll have to ask them whether they want me to just delete the tests or come up with a way to do on-device tests.

That's also how I ran into #2406.

That aside rebasing is trivial. The only merge conflict comes from a formatting change where `this._onNavigationStateChange(state.nav, nav, action));` was changed so `);` is on the next line.

**Comment by tlvenn** [comment]
Hi any chance someone can review this ? @matthamil I see that you are the currently assigned reviewer, do you have time to follow up on your previous review ?

Would be great to get this contribution merged, thanks in advance.

**Comment by dantman** [comment]
enzyme is a dead end, it won't actually work with React Native (https://github.com/airbnb/enzyme/issues/928) and the mock it depends on doesn't support react@16 either.

I looked more at `react-test-renderer` and it should be possible to get it to work at some point. However we're affected by some bugs in the test setup. We have to wait till [this commit](https://github.com/facebook/react-native/commit/165f87673002dec640fe86486296cd0d0847b333) is released in the next version of RN (v0.49.0) and we'll have to update react-navigation's devDeps/example to that. The other caveat is that the `ReactUpdates` mocking to mock the timers may need to be replaced with something equivalent for react@16.

**Comment by kelset** [comment]
Hey @dantman I'd love to take a shot and review this to merge it. It looks like there is a conflicting file, can you please update the PR?

**Comment by dantman** [comment]
@kelset The merge conflict is a minor whitespace change, the code is the same so the functionality is the same.

Is react-navigation going to upgrade the dev deps to react-native 0.49? I need that to be able to update the tests when I rebase. Changes in react-native between when I wrote the commit and now switched to versions of react-native where react's test renderer was broken.

**Comment by kelset** [comment]
Ok thanks for the clarification - yeah tbh I want to update the lib to the new versions of the deps; I'm a bit afraid something will break, tbh.

@spencercarli @skevy what do you think?

**Comment by kelset** [comment]
Hey @dantman we finally upgraded to 0.49 & flow 0.53 so I think this PR can be unlocked now, right?

**Comment by vonovak** [comment]
@dantman may I ask you to resolve the conflicts? Thank you.

**Comment by dantman** [comment]
It was tricky to get async setState working again for the tests, but I think I got it sorted out.

**Comment by vonovak** [comment]
Thanks @dantman for the write up in https://github.com/react-community/react-navigation/issues/1206#issuecomment-298149588 and the consequent PR. I tested this and it looks good to me so I'm (finally) merging. Thanks again!!

**Comment by kelset** [comment]
Awesome, thanks @dantman !!! 🚀

**Pull request merged**